pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,426 @@
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientidd4",
+			err:      nil,
+			pubsub:   true,
+			handler:  handler{false, "clientidd4", &msgChan},
+		},
+		{
+			desc:     "Unsubscribe from a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientidd4",
+			err:      nil,
+			pubsub:   false,
+			handler:  handler{false, "clientidd4", &msgChan},
+		},
+		{
+			desc:     "Unsubscribe from an already unsubscribed topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrNotSubscribed,
+			pubsub:   false,
+			handler:  handler{false, "clientid4", &msgChan},
+		},
+		{
+			desc:     "Unsubscribe from an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			pubsub:   false,
+			handler:  handler{false, "clientid4", &msgChan},
+		},
+		{
+			desc:     "Unsubscribe from a topic with empty ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			pubsub:   false,
+			handler:  handler{false, "", &msgChan},
+		},
+		{
+			desc:     "Subscribe to a new topic with an ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic+"2"),
+			clientID: "clientid55",
+			err:      nil,
+			pubsub:   true,
+			handler:  handler{true, "clientid5", &msgChan},
+		},
+		{
+			desc:     "Unsubscribe from a topic with an ID with failing handler",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic+"2"),
+			clientID: "clientid55",
+			err:      mqtt_pubsub.ErrFailedHandleMessage,
+			pubsub:   false,
+			handler:  handler{true, "clientid5", &msgChan},
+		},
+		{
+			desc:     "Subscribe to a new topic with subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic+"2", subtopic),
+			clientID: "clientid55",
+			err:      nil,
+			pubsub:   true,
+			handler:  handler{true, "clientid5", &msgChan},
+		},
+		{
+			desc:     "Unsubscribe from a topic with subtopic with an ID with failing handler",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic+"2", subtopic),
+			clientID: "clientid55",
+			err:      mqtt_pubsub.ErrFailedHandleMessage,
+			pubsub:   false,
+			handler:  handler{true, "clientid5", &msgChan},
+		},
+	}
+	for _, tc := range cases {
+		switch tc.pubsub {
+		case true:
+			err := pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+			assert.Equal(t, tc.err, err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, tc.err, err))
+		default:
+			err := pubsub.Unsubscribe(tc.clientID, tc.topic)
+			assert.Equal(t, tc.err, err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, tc.err, err))
+		}
+	}
+}
+
+type handler struct {
+	fail      bool
+	publisher string
+	msgChan   *chan (messaging.Message)
```

Comment:
Why pointer? What are reference types in Go?

Reply:
Reference types are maps channels and slices.
But I thought that since default is pass by value for all types (in general), that's why I used a channel pointer to be safe.
Searching for answer on google, will update it to just chan(messaging.Message) if possible!

Reply:
Changed it to non-pointer!